### PR TITLE
Add actions regex_replace and regex_name_lookup

### PR DESCRIPTION
**Pull request by drevell**
The regex_replace action replaces a regex match (or a subgroup thereof) with a given string. The replacement string can use subgroup references like $1 or ${groupname}, and can also use Go template expressions. An example is:

  - regex: '([a-zA-Z]+), (?P<adjective>[a-zA-Z]+) world!'
    subgroup: 1
    with: 'fresh'

This would transform a file containing "Hello, cool world!" to "Hello, fresh world!"

The regex_name_lookup action replaces every named subgroup with the template input variable of the same name. For example:

  - regex: 'Hello, (?P<adjective>[a-zA-Z]+) world!'

... and suppose the CLI user provided --var=adjective=fresh. This would also transform a file containing "Hello, cool world!" to "Hello, fresh world!"

### CODE REVIEW

File: templates/commands/render.go
Comment (sethvargo) [inline templates/commands/render.go:366]: Nit: it's not clear what `s` is unless you read the code - it looks like it's the Go Template?

File: templates/commands/render_action.go
Comment (sethvargo) [inline templates/commands/render_action.go:94]: Do we want to return a multierror here instead? It could be useful to get _all_ your errors at once instead of "fix error, run again, fix different error, run again, fix different error ..."

```suggestion
		templated, err := parseAndExecuteGoTmpl(re.Pos, re.Val, inputs)
		if err != nil {
		  merr = multierror.Append(merr, err)
			continue
		}

		compiled[i], err = regexp.Compile(templated)
		if err != nil {
			merr = multierror.Append(merr, model.ErrWithPos(re.Pos, "failed compiling regex: %w", err))
			continue
		}
```

File: templates/commands/render_action.go
Comment (sethvargo) [inline templates/commands/render_action.go:103]: Why would we do this vs iterating over the slice in reverse? It feels unnecessarily expensive.

File: templates/commands/render_action_regexnamelookup.go
Comment (sethvargo) [inline templates/commands/render_action_regexnamelookup.go:78]: Instead of reversing on L54 and then doing this, I'm not sure I understand why we can't walk the slice in reverse:

```go
for i := len(allMatches)-1; i >= 0; i-- {
  oneMatch := allMatches[i]
}
```

File: templates/commands/render_action_regexnamelookup_test.go
Comment (sethvargo) [inline templates/commands/render_action_regexnamelookup_test.go:98]: Super nit: I'm not sure "should_work" is helpful in the name.

File: templates/commands/render_action_regexreplace.go
Comment (sethvargo) [inline templates/commands/render_action_regexreplace.go:74]: I'm not sure I understand why we're returning the last replacement?

File: templates/commands/render.go
Comment (drevell) [reply templates/commands/render.go]: I did some variable renaming to make this more clear.

File: templates/commands/render_action.go
Comment (drevell) [reply templates/commands/render_action.go]: SGTM, done.

File: templates/commands/render_action.go
Comment (drevell) [reply templates/commands/render_action.go]: Because `reverse()` seemed more readable me. It lets us use `for _ oneMatch := range allMatches` and avoids creating more index variables. This code already has a confusing number of indices (FindAllSubmatchIndex returns a slice-of-slices-of-indices 😱 ), so avoiding more indices seemed desirable.

 I take it you disagree though; so I switched over to iterating in reverse. I don't have strong feelings on it.

File: templates/commands/render_action_regexnamelookup.go
Comment (drevell) [reply templates/commands/render_action_regexnamelookup.go]: Done

File: templates/commands/render_action_regexnamelookup_test.go
Comment (drevell) [reply templates/commands/render_action_regexnamelookup_test.go:98]: Let me send a follow PR to fix this, I've been doing this throughout this codebase.

File: templates/commands/render_action_regexreplace.go
Comment (drevell) [reply templates/commands/render_action_regexreplace.go:69]: The result of each transformation is pipelined into the following transformation, and the final result is returned. There's nothing special about the last replacement. Note the `=` rather than `:=` in `b, err = replaceWithTemplate(...`.

File: templates/commands/render_action_regexnamelookup.go
Comment (yolocs) [inline templates/commands/render_action_regexnamelookup.go:69]: Do we need `//nolint:wrapcheck` here?

File: templates/commands/render_action_regexreplace.go
Comment (sethvargo) [reply templates/commands/render_action_regexreplace.go:69]: Ah I missed the non-assignment. Thanks.

**Review by sethvargo — APPROVED** [review]

File: templates/commands/render_action_regexreplace.go
Comment (yolocs) [inline templates/commands/render_action_regexreplace.go:115]: This is some next level code complexity

**Review by yolocs — COMMENTED** [review]
Rant first - As a maintainer, while I appreciate the thorough documentation in code, I doubt whether I can effectively debug the regex related code in case something goes wrong. As a potential template builder myself, I would refuse to use regex because it would take me more time to debug whether I'm writing regex correctly.

---

Future - I'm interested in finding how our future users will use this. Would they prefer regex or the other ways?

---

For this PR - I don't have much Go readability comments - LGTM. But due to my limited regex knowledge, it's better to have some extra eyes to take a look.

File: templates/commands/render_action_regexnamelookup.go
Comment (drevell) [reply templates/commands/render_action_regexnamelookup.go:69]: Yep. The linter will flag any superfluous `//nolint` comments, so you can know that it's necessary.

**Comment by drevell** [comment]
> Rant first - As a maintainer, while I appreciate the thorough documentation in code, I doubt whether I can effectively debug the regex related code in case something goes wrong. As a potential template builder myself, I would refuse to use regex because it would take me more time to debug whether I'm writing regex correctly.
> 
> Future - I'm interested in finding how our future users will use this. Would they prefer regex or the other ways?
> 
> For this PR - I don't have much Go readability comments - LGTM. But due to my limited regex knowledge, it's better to have some extra eyes to take a look.

When you're back from OOO, let's get together and explore ways to simplify.

**Pull request merged**
